File: sdk/samples/WCFSamples/TechnologySamples/Scenario/WorkflowServices/DuplexWorkflowService/VB/ServiceWorkflow/ServiceWorkflow.xoml.vb
' Copyright (c) Microsoft Corporation. All rights reserved.

Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Collections.Generic
Imports System.Drawing
Imports System.ServiceModel
Imports System.Text
Imports System.Workflow.ComponentModel.Compiler
Imports System.Workflow.ComponentModel.Serialization
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.Runtime
Imports System.Workflow.Activities
Imports System.Workflow.Activities.Rules

Namespace Microsoft.WorkflowServices.Samples
    <Serializable()> _
    Partial Public Class ServiceWorkflow
        Inherits StateMachineWorkflowActivity
        Public Const ReverseEndpoint As String = "ReverseEndpoint"

        ' input of IForwardContract.BeginWorkflow
        <NonSerialized()> _
        Public ReturnAddress As EndpointAddress10 = Nothing

        ' input of IForwardContract.BeginWorkItem
        <NonSerialized()> _
        Public BeginWorkItemInput As String = Nothing

        ' input of IForwardContract.ContinueWorkItem
        <NonSerialized()> _
        Public ContinueWorkItemInput As Integer = Nothing

        ' input of IFowardContract.CompleteWorkItem
        <NonSerialized()> _
        Public CompleteWorkItemInput As String = Nothing

        Public CurrentWorkItem As WorkItem = Nothing
        Public CompletedWorkItems As New List(Of WorkItem)(3)

        Private Sub ApplyReturnAddress(ByVal sender As Object, ByVal e As EventArgs)
            ' apply ReturnAddress to ReverseEndpoint
            Dim epr As EndpointAddress = ReturnAddress.ToEndpointAddress()
            ContextManager.ApplyEndpointAddress(CType(GetActivityByName("SendWorkItemComplete"), SendActivity), epr)

            Dim context As Dictionary(Of String, String) = epr.Headers(0).GetValue(Of Dictionary(Of String, String))()
            Dim contextString As StringBuilder = New StringBuilder()

            For Each pair As KeyValuePair(Of String, String) In context
                contextString.Append(pair.Key)
                contextString.Append(":")
                contextString.Append(pair.Value)
            Next

            DebugOutput("[ServiceWorkflow:ApplyReturnAddress] " + contextString.ToString())

        End Sub

        Private Sub CreateWorkItem(ByVal sender As Object, ByVal e As EventArgs)
            ' create a new CurrentWorkItem
            CurrentWorkItem = New WorkItem()
            CurrentWorkItem.FirstPart = BeginWorkItemInput
            DebugOutput("[ServiceWorkflow:CreateWorkItem] " + BeginWorkItemInput)
        End Sub

        Private Sub AddPartsToWorkItem(ByVal sender As Object, ByVal e As EventArgs)
            ' add parts to CurrentWorkItem
            If CurrentWorkItem.PartsList Is Nothing Then
                CurrentWorkItem.PartsList = New List(Of Integer)()
            End If
            CurrentWorkItem.PartsList.Add(ContinueWorkItemInput)
            DebugOutput("[ServiceWorkflow:AddPartsToWorkItem] " + CType(ContinueWorkItemInput, String))
        End Sub

        Private Sub AddCompletedWorkItem(ByVal sender As Object, ByVal e As EventArgs)
            ' complete CurrentWorkItem
            CurrentWorkItem.LastPart = CompleteWorkItemInput
            CompletedWorkItems.Add(CurrentWorkItem)
            DebugOutput("[ServiceWorkflow:AddCompletedWorkItem] " + CompleteWorkItemInput)
        End Sub

        Private Sub RemoveCompletedWorkItem(ByVal sender As Object, ByVal e As EventArgs)
            ' remove completed WorkItem
            Console.ForegroundColor = ConsoleColor.Green
            Console.WriteLine("WorkItemComplete: " + CompletedWorkItems(0).FirstPart)
            Console.ResetColor()
            CompletedWorkItems.RemoveAt(0)
            DebugOutput("[ServiceWorkflow:RemoveCompletedWorkItem] " + CompletedWorkItems.Count.ToString())
        End Sub

        Private Sub DebugOutput(ByVal output As String)
            Console.WriteLine(output)
        End Sub

    End Class

End Namespace

File: sdk/samples/WCFSamples/TechnologySamples/Scenario/WorkflowServices/DuplexWorkflowService/CS/ServiceWorkflow/ServiceWorkflow.xoml.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Collections.Generic;
using System.Drawing;
using System.ServiceModel;
using System.Text;
using System.Workflow.ComponentModel.Compiler;
using System.Workflow.ComponentModel.Serialization;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Design;
using System.Workflow.Runtime;
using System.Workflow.Activities;
using System.Workflow.Activities.Rules;

namespace Microsoft.WorkflowServices.Samples
{
	public partial class ServiceWorkflow: StateMachineWorkflowActivity
	{
        public const string ReverseEndpoint = "ReverseEndpoint";
        [NonSerialized] // input of IForwardContract.BeginWorkflow
        public EndpointAddress10 ReturnAddress = default(EndpointAddress10);
        [NonSerialized] // input of IForwardContract.BeginWorkItem
        public string BeginWorkItemInput = default(System.String);
        [NonSerialized] // input of IForwardContract.ContinueWorkItem
        public int ContinueWorkItemInput = default(System.Int32);
        [NonSerialized] // input of IFowardContract.CompleteWorkItem
        public string CompleteWorkItemInput = default(System.String);

        public WorkItem CurrentWorkItem = null;
        public List<WorkItem> CompletedWorkItems = new List<WorkItem>(3);

        private void ApplyReturnAddress(object sender, EventArgs e)
        {   // apply ReturnAddress to ReverseEndpoint
            EndpointAddress epr = ReturnAddress.ToEndpointAddress();
            ContextManager.ApplyEndpointAddress(this.SendWorkItemComplete, epr);
            Dictionary<string, string> context = epr.Headers[0].GetValue<Dictionary<string,string>>();

            StringBuilder contextString = new StringBuilder();

            foreach(KeyValuePair<string, string> pair in context)
            {
                contextString.Append(pair.Key);
                contextString.Append(":");
                contextString.Append(pair.Value);
            }
            DebugOutput("[ServiceWorkflow:ApplyReturnAddress] " + contextString.ToString());
            
        }

        private void CreateWorkItem(object sender, EventArgs e)
        {   // create a new CurrentWorkItem
            CurrentWorkItem = new WorkItem();
            CurrentWorkItem.FirstPart = BeginWorkItemInput;
            DebugOutput("[ServiceWorkflow:CreateWorkItem] " + BeginWorkItemInput);
        }

        private void AddPartsToWorkItem(object sender, EventArgs e)
        {   // add parts to CurrentWorkItem
            if (CurrentWorkItem.PartsList == null)
                CurrentWorkItem.PartsList = new List<int>();
            CurrentWorkItem.PartsList.Add(ContinueWorkItemInput);
            DebugOutput("[ServiceWorkflow:AddPartsToWorkItem] " + ContinueWorkItemInput);
        }

        private void AddCompletedWorkItem(object sender, EventArgs e)
        {   // complete CurrentWorkItem
            CurrentWorkItem.LastPart = CompleteWorkItemInput;
            CompletedWorkItems.Add(CurrentWorkItem);
            DebugOutput("[ServiceWorkflow:AddCompletedWorkItem] " + CompleteWorkItemInput);
        }

        private void RemoveCompletedWorkItem(object sender, EventArgs e)
        {   // remove completed WorkItem
            Console.ForegroundColor = ConsoleColor.Green;
            Console.WriteLine("WorkItemComplete: " + CompletedWorkItems[0].FirstPart);
            Console.ResetColor();
            CompletedWorkItems.RemoveAt(0);
            DebugOutput("[ServiceWorkflow:RemoveCompletedWorkItem] " + CompletedWorkItems.Count.ToString());
        }

        private void DebugOutput(string output)
        {
            Console.WriteLine(output);
        }

    }

}
